### ISSUE TITLE
2 seconds delay is seen with Page navigation using WebView2

### ISSUE THREAD

**Issue by sowrab1994** [MicrosoftEdge/WebView2Feedback#2381]
**Description**
When we load a local page using SetVirtualHostNameToFolderMapping and the domain is non-existent we see a 2 secondsdelay in the page navigation to complete. 

**Version**
SDK: 1.0.1185.39
Runtime: 100.0.1185.44
Framework: WPF
OS: Win10

**Repro Steps**
1) Create a sample WPF Application
2) Create WebView2 instance and add it to a WPF grid. 
3) Set CoreWebView2.SetVirtualHostNameToFolderMapping to a virtual host name and folder consisting of sample HTML page.  
4) Register NavigationStarting and NavigationCompleted events and log the timestamps. 
4) Load the sample HTML Page inside the WebView.
5) Run the application. 

Note that the page loads after 2 seconds delay. That can be seen by taking the difference in timestamps logged in NavigationCompleted and NavigationStarting events. 

**Workaround**
If we add a host file entry mapping valid IP address (machine's IP) and virtual hostname (provided in SetVirtualHostNameToFolderMapping), the page loads fine without any delay. 

**Additional context**
It looks like it is trying to reach the host and causes the timeout.

**Comment by david-risney** [comment]
Yes it looks like there's still a part of the code trying to resolve the hostname. Good bug thanks!

If you pick a name that causes hostname resolution to just wait and timeout after 2s you'll see that when navigating to a SetVirtualHostNameToFolderMapping URI. For example, if you use app.local or anything .local you'll see this behavior. As a workaround if you pick the special TLD  `*.example`, like `app.example` hostname resolution should go very quick without hitting the network and no 2s timeout issue.

**Comment by VAllens** [comment]
> Couple of other points:
> 
> * We can always see DOMContentLoaded fires exactly after around 2s (in dev tools using performance.getEntries())
> * If we load directly using file:/// path it is always quick. But we want to load using virtualhost so that we can take advantage of all Web APIs, some of which are not available with file:/// origin

Thank you so much. :)

I spent **4 hours** using `Edge DevTools` to study why the `DOMContentLoaded` event is always triggered after a delay of 2 seconds. It drives me crazy!

`file://` can solve my problem !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
My issue is the same as that raised here.

But i did not use `SetVirtualHostNameToFolderMapping`.
I use `AddWebResourceRequestedFilter` and `WebResourceRequested`. 
After getting the request, I get the stream from the assembly embedded resource and respond to it.

**Comment by VAllens** [comment]
After changing to `file://`, the delay is shortened from 2 seconds to 16 milliseconds, which is a huge improvement.
![image](https://user-images.githubusercontent.com/5094922/188798162-0e6035e6-e54a-45dc-b78b-9acf78b774cd.png)

**Comment by david-risney** [comment]
The SetVirtualHostName and file: URIs are handled by the browser process. The WebResourceRequested event raises its event to you on the WebView2 UI thread in your host app. This means the WebResourceRequested event will be potentially much slower because it has a second process hop and it has to wait for the UI thread to be available. If all you need is to access local files, SetVirtualHostName or file: URIs are a better option than WebResourceRequested. But if you need the flexibility of providing your own custom stream then WebResourceRequested is required.

**Comment by VAllens** [comment]
Yes, I embed all static resources(html, js, css, png, etc...) into assembly resources. 
So I need to custom handle requests and responses in `WebResourceRequested` event.

**Comment by bzuillsmith** [comment]
Thank you @VAllens  for the suggested workaround using `file://`. I'm going to try that right now. Just so no one is confused, it's not an issue of `WebResourceRequested` taking a long time because of interprocess communication that's happening. It's also not an issue of the initial load of the control. When using the `WebResourceRequested` event, the browser shows the file is downloaded in very few ms (the interprocess communication is done). Then it waits until the 2s mark to actually render. There were no other indicators in the dev tools as to what was taking so long.

Also note that using NavigateToString is extremely fast and I can even load the entire google.com search page in 1.2s. It's only the WebResourceRequested approach that has this issue, possibly only when using a real looking http(s) url.

**Comment by SteveJobsCousin** [comment]
i'm facing the same slow loading when using any **hostName** rather then "app.example", despite the issue was reported in 2022 why it still exist till today in 2024 ?

**Comment by VAllens** [comment]
Is there any progress on this issue?

**Comment by jochenkirstaetter** [comment]
Any updates on this issue?

**Comment by xland** [comment]
The `file://` protocol doesn't solve my issue.

Whether I use `file://`,

or `SetVirtualHostNameToFolderMapping` with `webview->Navigate(L"https://dom.example/index.html")`,

or load the `Google` homepage,

or `webview->NavigateToString(L"<html><body>Hello, World!</body></html>");`

it always takes 2–4 seconds.

The window is created first, and after 2–4 seconds the page is rendered in the window.

Once the first window is successfully created, creating the second window is much faster.

My test environment:

WebView2: 1.0.3240.44
OS: Windows 11
CPU: i9 13900HX
Memory: 48GB

I tested two official Microsoft samples: `https://github.com/MicrosoftEdge/WebView2Samples`, and they exhibit the same issue.


These codes offer no assistance at all:
```c++
auto options = WRL::Make<CoreWebView2EnvironmentOptions>();
options->put_AdditionalBrowserArguments(L"--disable-features=msSmartScreenProtection");
options->put_AdditionalBrowserArguments(L"--no-proxy-server");
```

```c++
auto webView3 = webview.try_query<ICoreWebView2_3>();
webView3->SetVirtualHostNameToFolderMapping(localDomain.data(),L"UI",COREWEBVIEW2_HOST_RESOURCE_ACCESS_KIND_ALLOW);
auto url = std::format(L"https://{}/index.html", localDomain);
//url = L"file://D:/project/HorseJs/x64/Release/UI/index.html";
//url = L"https://www.baidu.com/";
webview->Navigate(url.data());
```

**Comment by xland** [comment]
I tested the same program on another computer, and the page loaded in under 1 second.

The computer:

- OS: Windows 10
- CPU: i9-9900K
- Memory: 32GB

Is this problem possibly caused by a recent Windows 11 update?
@david-risney  @champnic

**Comment by omascia** [comment]
### Instant Virtual Host Mapping in WebView2 — DNS Delay Workaround

If you're seeing **2–3 second delays** when using `SetVirtualHostNameToFolderMapping` in WebView2, here's a workaround that drastically reduces navigation time:

> ✅ Use a virtual hostname with an *empty fragment*, like:
>
> - `https://assets../page.html`
> - `https://./page.html`

These forms of hostnames trigger **immediate NXDOMAIN responses** in the local DNS resolver, which lets WebView2 fall through to the virtual host mapping **almost instantly**. This removes the usual DNS lookup delay caused by fake hostnames that don't resolve.

By contrast, common choices like `assets`, `pages.local`, or `internal.site` will incur full DNS lookup delays before WebView2 even attempts to apply your mapping.

### ✍️ Suggestion to the WebView2 Team

Ideally, WebView2 should:

> **Check** `SetVirtualHostNameToFolderMapping` **before attempting DNS resolution.**

This would:

- Eliminate the need for developers to craft “DNS-failing” hostnames
- Align virtual host behavior with performance expectations
- Avoid the counterintuitive UX of `.local` domains being slower than `.`

Right now, this clever hostname trick is more necessity than quirk — but it shouldn't have to be. The virtual hosting system is an elegant idea, trapped behind an awkward implementation detail.

(Copilot helped me make my words clearer: english is not my native language).
Yours,

**Comment by xland** [comment]
Using domains like `.example` or `.localhost` can also bypass DNS checks, for example:
"https://dom.example/index.html"
However, this does not solve all the 2~4 second delay issues.

**Comment by Burakuslendera** [comment]
## Same delay without folder mapping, and a NetLog capture naming the span

We hit this on a host that does not use `SetVirtualHostNameToFolderMapping` at all. It serves every asset itself from `AddWebResourceRequestedFilter` plus `put_Response`, in process, so nothing needs the synthetic host resolved. The delay is identical, and a NetLog capture names it:

```
BEGIN  HOST_RESOLVER_MANAGER_JOB  host="app.local:443"  types=[A,AAAA]
END    HOST_RESOLVER_MANAGER_JOB
                                   2.007 s
```

Application log for the same navigation: main document handed to `put_Response` at T, first subresource requested at T+2.012 s. The resolver job covers that interval exactly. Runtime 150.0.4078.83.

So the title is narrower than the bug. Any synthetic origin whose name does not resolve pays this, however the content is served.

## The `*.example` suggestion does not work

Worth saying plainly, because it is the advice this thread currently offers. We measured it by renaming the host to `.test`: 2.027 s, no change against a 2.03 s baseline. That matches what @Jimex reported on #3398 in 2023 and what @xland reported here in 2025.

The reason is that `.example`, `.test` and `.invalid` are reserved by RFC 2606 only in the sense that nobody may register them. The resolver still queries the network and still waits for the answer.

## What does work: a name under `.localhost`

RFC 6761 reserves `.localhost` as always resolving to the loopback interface, and Chromium answers it without a network query at all. Same machine, same frontend, only the host name changed:

| | `app.local` | `app.localhost` |
| --- | --- | --- |
| document served to first subresource | 2.012 to 2.041 s, seven runs | **11 to 79 ms** |
| launch to window visible | 2419 to 2543 ms | **495 to 508 ms** |

It also fixed a second symptom we had filed separately: in-origin navigations aborting with `ERR_CONNECTION_ABORTED` and never committing. 45 consecutive aborts before, 16 of 16 committing after. The two second window was where that race lived.

## One dead end worth recording

`--host-resolver-rules=MAP <host> 127.0.0.1` through `AdditionalBrowserArguments` does nothing, and the reason is not that the mapping fails. The switch never applies: `127.0.0.1` appears nowhere in a NetLog capture taken with it set. [`get_AdditionalBrowserArguments`](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions) documents that WebView2 silently ignores switches it blocks or cannot parse, without listing which ones. So a probe built on a Chromium flag cannot distinguish "not the cause" from "never ran", and two runs were spent learning only that.

Also ruled out by measurement, in case they come up: declaring `Content-Length`, holding the COM response and its `IStream` past `put_Response`, removing every `AddScriptToExecuteOnDocumentCreated` registration, and `--no-proxy-server`. Proxy auto-config was the leading suspect until the capture showed the `wpad:80` lookup failing in about 10 ms and the proxy resolving to `DIRECT` before the document was even served.

## Docs

The page this bites people on says nothing about choosing a host name, and its own example maps `demo`. Opened MicrosoftDocs/edge-developer#3880 to add a short subsection under both approaches.

Numbers are from one machine; the shape reproduced across seven runs.

**Comment by mikehoffms** [comment]
[PR 3880](https://github.com/MicrosoftDocs/edge-developer/pull/3880) (Document how the virtual host name affects navigation time)
https://github.com/MicrosoftDocs/edge-developer/pull/3880
